### PR TITLE
Add mli files for mlg files, remove unused values

### PR DESCRIPTION
**Pull request by SkySkimmer**
Also add promote-until-clean to mlg dune rules for easier debug (may
be incomplete for the git tracked dune files)

### CODE REVIEW

File: vernac/g_proofs.mli [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2018       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+val hint : Hints.hints_expr Pcoq.Entry.t
+(** Used byg G_ltac to add Hint Extern. *)
```
Comment (vbgl) [inline vernac/g_proofs.mli]: byg?

**Review by vbgl — APPROVED** [review]

**Comment by SkySkimmer** [comment]
you saw nothing

Gaëtan Gilbert

**Review by ejgallego — CHANGES_REQUESTED** [review]
This looks wrong, what's the motivation?

At least the complete set of witnesses for AST extension should be exported for each file; the best way indeed would be to have coqpp generating the `.mli`, but not having a header is just fine too.

File: toplevel/dune [hunk 2]
```diff
@@ -9,5 +9,6 @@
 
 (rule
  (targets g_toplevel.ml)
+ (mode promote-until-clean)
```
Comment (ejgallego) [inline toplevel/dune]: This also seems like a mess to me, it won't be long before people do commit these files by mistake.

Instead Coq should switch to a supported pre-processing solution in the OCaml ecosystem; however we are not the only ones having the problem so I guess a bit more discussion in the next Dune meeting will be good.

**Comment by SkySkimmer** [comment]
>This looks wrong, what's the motivation?

Build parallelism: every time something in eg G_rewrite is changed everyone who does `open Ltac_plugin` gets recompiled AFAICT
Being able to tell when we have some unused value in a mlg is also nice.

**Comment by ejgallego** [comment]
> Build parallelism: every time something in eg G_rewrite is changed everyone who does `open Ltac_plugin` gets recompiled AFAICT

Only if the digest of the `.cmi` file changes.

> Being able to tell when we have some unused value in a mlg is also nice.

The problem is that unused values in `mlg` files is a bit tricky, as such files declare AST extensions all the `wit_` functions _must_ be exported as they are the keys for the simulated extensinble `type` of the Coq's AST. IMHO we are not ready to make the AST private, and would we like to do so we should start with `constr_expr`.

Having `coqpp` produce a `.mli` file with such witnesses is something I have thought from time to time, in this case `coqpp` could produce `g_groung_wit.ml` and then the `.mli` would include it; but I dunno if that's robust enough, see for example https://discuss.ocaml.org/t/jane-streets-intf-ml-usage/3848/3

**Comment by ppedrot** [comment]
> all the wit_ functions must be exported as they are the keys for the simulated extensinble type of the Coq's AST

I don't agree at all. Most of the dynamic types defined in mlg files serve a local purpose, and a such should *not* be exported. Unfortunately we're tied to the EXTEND macros, but in an ideal world dynamic types that ought to be exposed in the API should be defined somewhere else and only referred to in the syntax files.

**Comment by ejgallego** [comment]
> Most of the dynamic types defined in mlg files serve a local purpose, and a such should _not_ be exported.

Can you provide an example of such a type?

> dynamic types that ought to be exposed in the API

How do you infer the set of types that should be exposed?

[hint: what you propose does break all SerAPI users for example, I get the impression that any user case that is not good for you is not good for everyone else]

**Comment by ppedrot** [comment]
> Can you provide an example of such a type?

All the files that were given empty mli by @SkySkimmer demonstrate this. The arguments defined there are purely used for the tactic / vernac extensions.

> what you propose does break all SerAPI users

That might be true, and I already exposed my view, but let me reiterate here. SerAPI by its very nature of a serialization library has export requirements that go straightly against the very concept of abstraction. The root of the issue seems to be linked to the engineering choices of the OCaml language. Instead of exporting internal data structures all over the place, SerAPI should be part of the core Coq implementation and types should be exported only via a serialization / deserialization pair.

**Comment by ejgallego** [comment]
What is this concept of "abstraction"?

This PR makes the AST abstract, no sane language has an abstract AST [even tho OCaml is proposing to have one now] as abstraction buys you nothing. Clients need the AST, [not only for serialization but for other uses such as plugins], so even if you abtract you'll have to re-expose it again.

If you want to make the AST abstract then please start with `constr_expr` first.

> SerAPI should be part of the core Coq implementation and types should be exported only via a serialization / deserialization pair.

This goes against modularity, I do prefer modularity in this case over abstraction which is not abstraction.

> All the files that were given empty mli by @SkySkimmer demonstrate this. The arguments defined there are purely used for the tactic / vernac extensions.

This is just a coincidence, but note how there are arbitrary exports becuase you otherwise break valid use cases. This also goes strongly against #9547 .

I do agree that the current system should be better, but this is not the way forward.

We have people doing research in machine learning having to cut and paste code and use object magic because we broke things **for no gain at all**, this is a strong disservice to users and the community.

File: toplevel/dune (hunk 2 above)
Comment (Zimmi48) [reply toplevel/dune]: The fact that there is a rule `g_*.ml` in `.gitignore` should prevent this.

**Comment by Zimmi48** [comment]
> We have people doing research in machine learning having to cut and paste code and use `Obj.magic` because we broke things **for no gain at all**, this is a strong disservice to users and the community.

I am assuming you are referring to what's discussed in https://coq.discourse.group/t/machine-learning-and-hammers-for-coq/303?

**Comment by SkySkimmer** [comment]
>Only if the digest of the .cmi file changes.

Test: `make -f Makefile.dune check`, then pick a boolean in g_rewrite.mlg and toggle it (eg the one in `  | [ "<-" constr(c) ] -> { StratConstr (c, false) }`)
All of this gets recompiled:
~~~
dune build  @check
       coqpp plugins/ltac/g_rewrite.ml
    ocamldep plugins/ltac/.ltac_plugin.objs/g_rewrite.ml.d
      ocamlc plugins/ltac/.ltac_plugin.objs/byte/ltac_plugin__G_rewrite.{cmi,cmo,cmt}
      ocamlc doc/plugin_tutorial/tuto0/src/.tuto0_plugin.objs/byte/tuto0_plugin__Tuto0_main.{cmi,cmti}
      ocamlc doc/plugin_tutorial/tuto1/src/.tuto1_plugin.objs/byte/tuto1_plugin__Simple_check.{cmi,cmti}
      ocamlc doc/plugin_tutorial/tuto1/src/.tuto1_plugin.objs/byte/tuto1_plugin__Simple_declare.{cmi,cmti}
      ocamlc doc/plugin_tutorial/tuto1/src/.tuto1_plugin.objs/byte/tuto1_plugin__Simple_print.{cmi,cmti}
      ocamlc doc/plugin_tutorial/tuto0/src/.tuto0_plugin.objs/byte/tuto0_plugin__Tuto0_main.{cmo,cmt}
      ocamlc doc/plugin_tutorial/tuto1/src/.tuto1_plugin.objs/byte/tuto1_plugin__Simple_check.{cmo,cmt}
      ocamlc doc/plugin_tutorial/tuto1/src/.tuto1_plugin.objs/byte/tuto1_plugin__Simple_print.{cmo,cmt}
      ocamlc doc/plugin_tutorial/tuto1/src/.tuto1_plugin.objs/byte/tuto1_plugin__Simple_declare.{cmo,cmt}
      ocamlc doc/plugin_tutorial/tuto3/src/.tuto3_plugin.objs/byte/tuto3_plugin__Construction_game.{cmi,cmti}
      ocamlc doc/plugin_tutorial/tuto3/src/.tuto3_plugin.objs/byte/tuto3_plugin__Tuto_tactic.{cmi,cmti}
      ocamlc plugins/btauto/.btauto_plugin.objs/byte/btauto_plugin__Refl_btauto.{cmi,cmti}
      ocamlc doc/plugin_tutorial/tuto2/src/.tuto2_plugin.objs/byte/tuto2_plugin__Demo.{cmi,cmo,cmt}
      ocamlc doc/plugin_tutorial/tuto3/src/.tuto3_plugin.objs/byte/tuto3_plugin__Construction_game.{cmo,cmt}
      ocamlc plugins/cc/.cc_plugin.objs/byte/cc_plugin__Ccalgo.{cmi,cmti}
      ocamlc doc/plugin_tutorial/tuto3/src/.tuto3_plugin.objs/byte/tuto3_plugin__Tuto_tactic.{cmo,cmt}
      ocamlc plugins/derive/.derive_plugin.objs/byte/derive_plugin__Derive.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Big.{cmi,cmo,cmt}
      ocamlc plugins/btauto/.btauto_plugin.objs/byte/btauto_plugin__Refl_btauto.{cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Miniml.{cmi,cmti}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Formula.{cmi,cmti}
      ocamlc plugins/derive/.derive_plugin.objs/byte/derive_plugin__Derive.{cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Miniml.{cmo,cmt}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Unify.{cmi,cmti}
      ocamlc plugins/ltac/.tauto_plugin.objs/byte/tauto_plugin__Tauto.{cmi,cmti}
      ocamlc plugins/cc/.cc_plugin.objs/byte/cc_plugin__Ccalgo.{cmo,cmt}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Formula.{cmo,cmt}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Unify.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__G_micromega.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Itv.{cmi,cmti}
      ocamlc plugins/ltac/.tauto_plugin.objs/byte/tauto_plugin__Tauto.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Persistent_cache.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Micromega.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Itv.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Sos_lib.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Sos_types.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Persistent_cache.{cmo,cmt}
      ocamlc plugins/nsatz/.nsatz_plugin.objs/byte/nsatz_plugin__Nsatz.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Sos_types.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Sos_lib.{cmo,cmt}
      ocamlc plugins/nsatz/.nsatz_plugin.objs/byte/nsatz_plugin__G_nsatz.{cmi,cmo,cmt}
      ocamlc plugins/nsatz/.nsatz_plugin.objs/byte/nsatz_plugin__Utile.{cmi,cmti}
      ocamlc plugins/nsatz/.nsatz_plugin.objs/byte/nsatz_plugin__Polynom.{cmi,cmti}
      ocamlc plugins/omega/.omega_plugin.objs/byte/omega_plugin__Coq_omega.{cmi,cmti}
      ocamlc plugins/nsatz/.nsatz_plugin.objs/byte/nsatz_plugin__Utile.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Micromega.{cmo,cmt}
      ocamlc plugins/rtauto/.rtauto_plugin.objs/byte/rtauto_plugin__Proof_search.{cmi,cmti}
      ocamlc plugins/omega/.omega_plugin.objs/byte/omega_plugin__G_omega.{cmi,cmo,cmt}
      ocamlc plugins/nsatz/.nsatz_plugin.objs/byte/nsatz_plugin__Polynom.{cmo,cmt}
      ocamlc plugins/setoid_ring/.newring_plugin.objs/byte/newring_plugin__Newring_ast.{cmi,cmti}
      ocamlc plugins/ssrmatching/.ssrmatching_plugin.objs/byte/ssrmatching_plugin__Ssrmatching.{cmi,cmti}
      ocamlc plugins/rtauto/.rtauto_plugin.objs/byte/rtauto_plugin__Proof_search.{cmo,cmt}
      ocamlc plugins/omega/.omega_plugin.objs/byte/omega_plugin__Omega.{cmi,cmo,cmt}
      ocamlc plugins/setoid_ring/.newring_plugin.objs/byte/newring_plugin__Newring_ast.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2dyn.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2match.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2dyn.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2stdlib.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2match.{cmo,cmt}
      ocamlc doc/plugin_tutorial/tuto0/src/.tuto0_plugin.objs/byte/tuto0_plugin__G_tuto0.{cmi,cmo,cmt}
      ocamlc dev/.top_printers.objs/byte/top_printers.{cmi,cmti}
      ocamlc doc/plugin_tutorial/tuto1/src/.tuto1_plugin.objs/byte/tuto1_plugin__G_tuto1.{cmi,cmo,cmt}
      ocamlc doc/plugin_tutorial/tuto3/src/.tuto3_plugin.objs/byte/tuto3_plugin__G_tuto3.{cmi,cmo,cmt}
      ocamlc dev/.top_printers.objs/byte/top_printers.{cmo,cmt}
      ocamlc plugins/btauto/.btauto_plugin.objs/byte/btauto_plugin__G_btauto.{cmi,cmo,cmt}
      ocamlc plugins/cc/.cc_plugin.objs/byte/cc_plugin__Ccproof.{cmi,cmti}
      ocamlc plugins/derive/.derive_plugin.objs/byte/derive_plugin__G_derive.{cmi,cmo,cmt}
      ocamlc plugins/ssrmatching/.ssrmatching_plugin.objs/byte/ssrmatching_plugin__Ssrmatching.{cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Common.{cmi,cmti}
      ocamlc plugins/omega/.omega_plugin.objs/byte/omega_plugin__Coq_omega.{cmo,cmt}
      ocamlc plugins/cc/.cc_plugin.objs/byte/cc_plugin__Ccproof.{cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Haskell.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Extract_env.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Extraction.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Json.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Modutil.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Ocaml.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Scheme.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Table.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Coq_micromega.{cmi,cmti}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Sequent.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Mutils.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__G_micromega.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Certificate.{cmi,cmti}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Sequent.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Sos.{cmi,cmti}
      ocamlc plugins/nsatz/.nsatz_plugin.objs/byte/nsatz_plugin__Ideal.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Mutils.{cmo,cmt}
      ocamlc plugins/rtauto/.rtauto_plugin.objs/byte/rtauto_plugin__Refl_tauto.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Table.{cmo,cmt}
      ocamlc plugins/setoid_ring/.newring_plugin.objs/byte/newring_plugin__Newring.{cmi,cmti}
      ocamlc plugins/nsatz/.nsatz_plugin.objs/byte/nsatz_plugin__Ideal.{cmo,cmt}
      ocamlc plugins/ssrmatching/.ssrmatching_plugin.objs/byte/ssrmatching_plugin__G_ssrmatching.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Sos.{cmo,cmt}
      ocamlc plugins/rtauto/.rtauto_plugin.objs/byte/rtauto_plugin__Refl_tauto.{cmo,cmt}
      ocamlc plugins/cc/.cc_plugin.objs/byte/cc_plugin__Cctac.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2expr.{cmi,cmti}
      ocamlc plugins/ssrmatching/.ssrmatching_plugin.objs/byte/ssrmatching_plugin__G_ssrmatching.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2core.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Mlutil.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__G_extraction.{cmi,cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Functional_principles_types.{cmi,cmti}
      ocamlc plugins/cc/.cc_plugin.objs/byte/cc_plugin__Cctac.{cmo,cmt}
      ocamlc plugins/setoid_ring/.newring_plugin.objs/byte/newring_plugin__Newring.{cmo,cmt}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Ground.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Vect.{cmi,cmti}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Rules.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Vect.{cmo,cmt}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__G_ground.{cmi,cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Mlutil.{cmo,cmt}
      ocamlc plugins/rtauto/.rtauto_plugin.objs/byte/rtauto_plugin__G_rtauto.{cmi,cmo,cmt}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Rules.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrvernac.{cmi,cmti}
      ocamlc plugins/cc/.cc_plugin.objs/byte/cc_plugin__G_congruence.{cmi,cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrast.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2ffi.{cmi,cmti}
      ocamlc plugins/setoid_ring/.newring_plugin.objs/byte/newring_plugin__G_newring.{cmi,cmo,cmt}
      ocamlc plugins/nsatz/.nsatz_plugin.objs/byte/nsatz_plugin__Nsatz.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrbwd.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2intern.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2qexpr.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2ffi.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2entries.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Haskell.{cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Common.{cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Json.{cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Extract_env.{cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Modutil.{cmo,cmt}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Scheme.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Glob_term_to_relation.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Extraction.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Glob_termops.{cmi,cmti}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Indfun.{cmi,cmti}
      ocamlc plugins/extraction/.extraction_plugin.objs/byte/extraction_plugin__Ocaml.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Invfun.{cmi,cmti}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Indfun_common.{cmi,cmti}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Instances.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Polynomial.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrcommon.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrelim.{cmi,cmti}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Instances.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrequality.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssripats.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrfwd.{cmi,cmti}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Indfun_common.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrprinters.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrtacticals.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Polynomial.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrview.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2env.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrprinters.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrtacticals.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2print.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2env.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2types.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrview.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2extffi.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2quote.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2extffi.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Recdef.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2print.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__G_indfun.{cmi,cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Functional_principles_proofs.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2quote.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Glob_termops.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Glob_term_to_relation.{cmo,cmt}
      ocamlc plugins/firstorder/.ground_plugin.objs/byte/ground_plugin__Ground.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Functional_principles_proofs.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Recdef.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Mfourier.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Simplex.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrparser.{cmi,cmti}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Simplex.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Invfun.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Mfourier.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrelim.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrequality.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrbwd.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrcommon.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2interp.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrfwd.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2interp.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2tactics.{cmi,cmti}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssripats.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2tactics.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2intern.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Functional_principles_types.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrparser.{cmo,cmt}
      ocamlc plugins/funind/.recdef_plugin.objs/byte/recdef_plugin__Indfun.{cmo,cmt}
      ocamlc plugins/micromega/.csdpcert.eobjs/byte/csdpcert.{cmi,cmti}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__G_ltac2.{cmi,cmo,cmt}
      ocamlc plugins/micromega/.csdpcert.eobjs/byte/csdpcert.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Certificate.{cmo,cmt}
      ocamlc plugins/micromega/.micromega_plugin.objs/byte/micromega_plugin__Coq_micromega.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2entries.{cmo,cmt}
      ocamlc plugins/ssr/.ssreflect_plugin.objs/byte/ssreflect_plugin__Ssrvernac.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2stdlib.{cmo,cmt}
      ocamlc user-contrib/Ltac2/.ltac2_plugin.objs/byte/ltac2_plugin__Tac2core.{cmo,cmt}
~~~

**Comment by ejgallego** [comment]
> All of this gets recompiled:

What are the digests of `ltac_plugin__G_rewrite.cmi` before and after.

**Comment by SkySkimmer** [comment]
Assuming digest means md5sum:
Before d9985e739c05fb8e1a350e119e6851f7
After 89e57fef0f25184f1c41db6a8d6812d4

**Comment by SkySkimmer** [comment]
In fact even if all I do is add an empty line somewhere it changes.

**Comment by ejgallego** [comment]
> In fact even if all I do is add an empty line somewhere it changes.

That's quite bizarre! Why is OCaml behaving like that?

**Comment by ejgallego** [comment]
> That's quite bizarre! Why is OCaml behaving like that?

What's the diff on the generated `.ml` files?

**Comment by SkySkimmer** [comment]
>Maybe coqpp is modifying some line annotations behind the back?

Of course it does, if you add a line somewhere all the annotations after it need to be bumped.

**Comment by ejgallego** [comment]
> Of course it does, if you add a line somewhere all the annotations after it need to be bumped.

Unfortunate :(

Anyways, unfortunately, the `.mli` files cannot be added like that, as they make the extended AST private breaking many users; an alternative solution should be found.

**Comment by ejgallego** [comment]
> Of course it does, if you add a line somewhere all the annotations after it need to be bumped.

Why does the `bool` change trigger a `cmi` digest difference tho?

**Comment by Zimmi48** [comment]
"many users" seems a bit excessive phrasing, especially without more evidence. Maybe you wanted to say "some invaluable and legitimate use cases".

**Comment by ejgallego** [comment]
No, I wanted to say "many users", including Princeton U., Urbana Champaign, U. Texas at Austin, MINES, UCSD, Galois Inc. Grammatech Inc. Eindhoven Univ. of Tech. all jsCoq users [on the dozens _easily_] , and many others I'm not up to date [MIT? etc...]

Lack of trust on fellow developers is staggering.

**Comment by SkySkimmer** [comment]
>Why does the bool change trigger a cmi digest difference tho?

I guess because `true` and `false` don't have the same length.

**Comment by ejgallego** [comment]
> I guess because `true` and `false` don't have the same length.

I see, thanks.

**Comment by ejgallego** [comment]
This needs discussion, if we merge this PR then there is no way to get the AST of Coq documents anymore, that seems serious to me.

**Comment by SkySkimmer** [comment]
Don't care enough to push farther.

**Pull request closed without merging**